pkg/network/aci_network.go

```diff
@@ -101,33 +111,60 @@ func (pn *ProviderNetwork) SetVNETConfig(ctx context.Context, azConfig *auth.Con
 
 func (pn *ProviderNetwork) setupNetwork(ctx context.Context, azConfig *auth.Config) error {
 	logger := log.G(ctx).WithField("method", "setupNetwork")
-	logger.Debug("setting up network")
+	ctx, span := trace.StartSpan(ctx, "network.setupNetwork")
+	defer span.End()
+
+	logger.Debug("getting azure credential")
 
-	c := aznetwork.NewSubnetsClient(azConfig.AuthConfig.SubscriptionID)
-	c.Authorizer = azConfig.Authorizer
+	var err error
+	var credential azcore.TokenCredential
+	isUserIdentity := len(azConfig.AuthConfig.ClientID) == 0
 
+	if isUserIdentity {
+		credential, err = azConfig.GetMSICredential(ctx)
+	} else {
+		credential, err = azConfig.GetSPCredential(ctx)
+	}
+	if err != nil {
+		return errors.Wrap(err, "an error has occurred while creating getting credential ")
+	}
+
+	options := arm.ClientOptions{
+		ClientOptions: azcore.ClientOptions{
+			Cloud: azConfig.Cloud,
+		},
+	}
+
+	var rawResponse *http.Response
+	ctxWithResp := runtime.WithCaptureResponse(ctx, &rawResponse)
+
+	subnetsClient, err := aznetworkv2.NewSubnetsClient(azConfig.AuthConfig.SubscriptionID, credential, &options)
+	if err != nil {
+		return errors.Wrap(err, "an error has occurred while creating subnet client")
+	}
 	createSubnet := true
-	subnet, err := c.Get(ctx, pn.VnetResourceGroup, pn.VnetName, pn.SubnetName, "")
+	response, err := subnetsClient.Get(ctx, pn.VnetResourceGroup, pn.VnetName, pn.SubnetName, nil)
 	if err != nil && !errdefs.IsNotFound(err) {
 		return fmt.Errorf("error while looking up subnet: %v", err)
 	}
 	if errdefs.IsNotFound(err) && pn.SubnetCIDR == "" {
 		return fmt.Errorf("subnet '%s' is not found in vnet '%s' in resource group '%s' and subscription '%s' and subnet CIDR is not specified", pn.SubnetName, pn.VnetName, pn.VnetResourceGroup, pn.VnetSubscriptionID)
 	}
 	if err == nil {
+		currentSubnet := response.Subnet
 		if pn.SubnetCIDR == "" {
```

Comment:
I suggest to add a nil check for currentSubnet.Properties.AddressPrefix

if currentSubnet.Properties.AddressPrefix != nil {

if pn.SubnetCIDR == "" {
    pn.SubnetCIDR = *currentSubnet.Properties.AddressPrefix
}

if pn.SubnetCIDR != *currentSubnet.Properties.AddressPrefix {
			   return fmt.Errorf("found subnet '%s' using different CIDR: '%s'. desired: '%s'", pn.SubnetName, 
    *currentSubnet.Properties.AddressPrefix, pn.SubnetCIDR)
		}
}